### PR TITLE
RF/CUMULUS-3861: Upgrade to React 18 and to React Testing Library

### PR DESCRIPTION
**Pull request by vindigo**
## Summary
Project: CUMULUS-3860: Dashboard: Upgrade from React 17.x to 18.x
Reason: Technical debt request.

Addresses [CUMULUS-3860: Dashboard: Upgrade React 17 to React 18.
https://bugs.earthdata.nasa.gov/browse/CUMULUS-3860

Project: CUMULUS-3861: Dashboard: Replace Enzyme with React Testing Library(RTL)
Reason: Removed references to Enzyme and replaced them with React compliant testing components.

Addresses [CUMULUS-3861: Dashboard: Replace Enzyme with React Testing Library(RTL)
https://bugs.earthdata.nasa.gov/browse/CUMULUS-3861

## Changes

* Upgrade from React 17.x to 18.x
* Upgraded existing Enzyme test to React Testing LIbrary
* All test assertions in the test scripts are now passing without reference to Enzyme. Total of about 29 test assertions among these test scripts.

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Adhoc testing
- [ ] Integration tests

### CODE REVIEW

File: test/components/table/TableFilters.js [hunk 1]
```diff
@@ -30,33 +27,34 @@ const columns = [
 ];
 
 test('Renders table filters', t => {
-  const wrapper = shallow(
+  const { container } = render(
     <TableFilters columns={columns} />
   );
 
-  const filters = wrapper.find('.table__filters--filter');
+  const filters = container.querySelectorAll('.table__filters--filter');
```
Comment (Jkovarik) [inline test/components/table/TableFilters.js:34]: (this comment needs to be prefaced that the team should think about if this is a block/debt or nit): 

These kinds of queries should *probably* be moved to https://testing-library.com/docs/queries/about/#queries 

e.g. https://testing-library.com/docs/marko-testing-library/api/#container-1 

tl;dr querySelector as a query strategy may be vulnerable to DOM changes/ https://medium.com/smallcase-engineering/why-should-we-not-use-document-queryselector-while-testing-react-components-835bbf34169e#7aa1

Comment (BWex-NASA) [reply test/components/table/TableFilters.js:34]: Thanks @Jkovarik. This test was converted by @Nnaga1, but @vindigo and I have used the `querySelector` in some of our tests as well. Some of our tests we were able to use the `screen.findByText` instead of `querySelector`. However, in this particular test I tried using the following method and it results in a failure:  `const filters = screen.getAllByText('.table__filters--filter');`

Perhaps @Nnaga1 might know the correct method to use.

Comment (BWex-NASA) [reply test/components/table/TableFilters.js:34]: Thanks @Nnaga1, this change was in reference to @Jkovarik first comment above: https://github.com/nasa/cumulus-dashboard/pull/1158#discussion_r1889256054

Comment (Jkovarik) [reply test/components/table/TableFilters.js:34]: @Nnaga1  see the original comment with links.        tl;dr raising the question because enzyme suggests it as a best practice.   See links in original post, and/or https://testing-library.com/docs/marko-testing-library/api/#container-1

File: README.md [hunk 2]
```diff
@@ -184,6 +184,12 @@ Otherwise, follow the instructions for building and deploying the dashboard for
   $ npm run test
 ```
 
+Command to run a single test script locally using Ava:
```
Comment (Jkovarik) [inline README.md:187]: 😍

File: localdashboard_cumulus_std.bash [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
```
Comment (Jkovarik) [inline localdashboard_cumulus_std.bash:1]: - Nit: consider .sh extension as other bash scripts in this repo are .bash
- Please consider moving to ./bin/
- Nit: add a brief description re: what this does
- Nit: consider adding this to either the readme or as an npm script to make it easier to find/awareness/etc.

File: localdashboard_cumulus_std.bash [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+# SIT shared dashboard
+export APIROOT=https://d84ilrgn1b7e1.cloudfront.net/
```
Comment (Jkovarik) [inline localdashboard_cumulus_std.bash:3]: This config value could change.   This may also be a security/obscurity concern.

File: localdashboard_cumulus_std.bash (hunk 3 above)
Comment (Jkovarik) [reply localdashboard_cumulus_std.bash:1]: Also: this appears to be a script setup to automate building against a particular SIT env - is it a best practice to do this versus a development stack or the local test stack?

File: test/components/granules/granule.js [hunk 5]
```diff
@@ -1,17 +1,20 @@
...
+import thunk from 'redux-thunk';
+import { initialState } from '../../../app/src/js/reducers/datepicker';
+import { Provider } from 'react-redux';
 
 import { GranuleOverview } from '../../../app/src/js/components/Granules/granule.js';
 
-configure({ adapter: new Adapter() });
-
-const logs = { items: [] };
+const logs = { items: [''] };
```
Comment (Jkovarik) [inline test/components/granules/granule.js:17]: Why was this change needed?

**Review by Jkovarik — CHANGES_REQUESTED** [review]
This looks really good, just a couple of questions/concerns to talk over!

File: test/components/table/TableFilters.js (hunk 1 above)
Comment (Jkovarik) [reply test/components/table/TableFilters.js:34]: @Nnaga1  we should probably post-scrum and chat further.

File: CHANGELOG.md [hunk 6]
```diff
@@ -6,6 +6,34 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Changed
+
+- **CUMULUS-3861**
+  - Updated React 17 to React 18
+  - Updated Enzyme tests to React Testing Library
```
Comment (jennyhliu) [inline CHANGELOG.md:13]: Does this already cover the next item "Removed test script references..."? It's not totally necessary to list the test scripts. Changelog only needs to summarize the changes.

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,34 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
...
+      - test/components/executions/execution-status.js
+      - test/components/granules/bulk-granule-modal.js
+      - test/components/header/condition.js
+
+### Changed
+
+- **CUMULUS-3870**
+  - Updated React and dependencies
+  - react 17.0.2 ⇒ 18.3.1
+  - react-ace 9.5.0 ⇒ 9.5.0
```
Comment (jennyhliu) [inline CHANGELOG.md:27]: react-ace is updated to ^13.0.0
We only need to call out the major version updates, the lines below can be removed.

File: CHANGELOG.md [hunk 8]
```diff
@@ -32,12 +60,26 @@ This version of the dashboard requires Cumulus API >= v19.2.0-alpha.1 (TBD API r
 - **CUMULUS-3905**
   - Fixed npm audit issues with `body-parser` and `path-to-regexp`
 
+### Changed
+- **CUMULUS-3860**
```
Comment (jennyhliu) [inline CHANGELOG.md:64]: What's the difference between CUMULUS-3860 and CUMULUS-3870? The change log under CUMULUS-3870 already stated the version updates `Updated React and dependencies`.

File: CHANGELOG.md [hunk 9]
```diff
@@ -32,12 +60,26 @@ This version of the dashboard requires Cumulus API >= v19.2.0-alpha.1 (TBD API r
...
+  - react-select 5.2.1 ⇒ 5.8.1
+  - react-table 7.6.1 ⇒ 7.6.1
+  - react-test-renderer 18.3.1 ⇒ 18.3.1
+
 ## [v12.2.0] - 2024-09-04
 
 This version of the dashboard requires Cumulus API >= v18.4.0
 
 ### Changed
-
```
Comment (jennyhliu) [inline CHANGELOG.md:40]: keep the blank line.

File: localdashboard_cumulus_std.bash
Comment (jennyhliu) [inline localdashboard_cumulus_std.bash:1]: This is for internal use, and shared in slack files, can be removed here.

**Comment by Jkovarik** [comment]
We should probably also add a CHANGELOG note/log re: the Bamboo changes, and before this gets merged make sure we've documented the dependency override and have an ADR.

File: test/components/collections/list.js [hunk 10]
```diff
@@ -49,60 +51,66 @@ const providers = {
...
 };
 
+const dispatch = () => {};
+const logs = {};
+const locationQueryParams = {
+  search: {}
+};
+
+const middlewares = [requestMiddleware, thunk];
+const mockStore = configureMockStore(middlewares);
```
[the remark this answers is not in the document]
Comment (jjmccoy) [reply test/components/collections/list.js:61]: As discussed, yes this convention is deprecated; however, the code involving the store will be updated in the Redux ticket: [https://bugs.earthdata.nasa.gov/browse/CUMULUS-3863](CUMULUS-3863).

File: test/components/granules/bulk-granule-modal.js [hunk 11]
```diff
@@ -1,107 +1,111 @@
...
+import thunk from 'redux-thunk';
+import configureMockStore from 'redux-mock-store';
 
-import { BulkGranuleModal } from '../../../app/src/js/components/Granules/bulk-granule-modal.js';
-
-configure({ adapter: new Adapter() });
+const locationQueryParams = { search: {} };
+const dispatch = () => {};
+const middlewares = [requestMiddleware, thunk];
+const mockStore = configureMockStore(middlewares);
```
[the remark this answers is not in the document]
Comment (jjmccoy) [reply test/components/granules/bulk-granule-modal.js:15]: Please see the answer to previous PR comment concerning the store.

File: test/components/executions/execution-events_correctStepName.js
Comment (jennyhliu) [inline test/components/executions/execution-events_correctStepName.js:1]: Why do you rename execution-events.js to execution-events_correctStepName.js?
execution-events.js corresponds to the components/execution-events.js

File: app/src/js/components/Granules/test.container.html
Comment (jennyhliu) [inline app/src/js/components/Granules/test.container.html:1]: What's this file for?

File: package.json [hunk 12]
```diff
@@ -143,27 +144,31 @@
     "@fortawesome/free-solid-svg-icons": "^5.15.4",
     "@fortawesome/react-fontawesome": "^0.1.16",
     "@reduxjs/toolkit": "^1.7.1",
-    "bootstrap": "^4.5.3",
+    "@tanstack/react-table": "^8.20.5",
+    "@testing-library/dom": "^10.4.0",
```
Comment (jennyhliu) [inline package.json:148]: Where is it used? Even if it's used, I think it should be under `devDependencies`

File: package.json [hunk 13]
```diff
@@ -71,7 +71,8 @@
     "@cumulus/api": "19.2.0-alpha.1",
     "@cumulus/aws-client": "19.2.0-alpha.1",
     "@cypress/webpack-preprocessor": "^5.11.0",
-    "@wojtekmaj/enzyme-adapter-react-17": "^0.6.6",
+    "@testing-library/jest-dom": "^6.5.0",
```
Comment (jennyhliu) [inline package.json:74]: Where do we need these @testing-library/ libraries?

File: package.json [hunk 14]
```diff
@@ -105,6 +105,7 @@
     "eslint-plugin-react": "^7.28.0",
     "eslint-plugin-react-hooks": "^4.3.0",
     "eslint-plugin-standard": "^5.0.0",
+    "eslint-plugin-testing-library": "^6.3.0",
```
Comment (jennyhliu) [inline package.json:108]: Where is this used?

File: package.json [hunk 15]
```diff
@@ -143,27 +144,31 @@
     "@fortawesome/free-solid-svg-icons": "^5.15.4",
     "@fortawesome/react-fontawesome": "^0.1.16",
     "@reduxjs/toolkit": "^1.7.1",
-    "bootstrap": "^4.5.3",
+    "@tanstack/react-table": "^8.20.5",
```
Comment (jennyhliu) [inline package.json:147]: Where do we need this? We use "react-table"

File: package.json [hunk 16]
```diff
@@ -143,27 +144,31 @@
...
+    "@testing-library/dom": "^10.4.0",
+    "ace-builds": "^1.36.2",
```
Comment (jennyhliu) [inline package.json:149]: Where is ace-builds used?

File: package.json [hunk 17]
```diff
@@ -143,27 +144,31 @@
...
-    "final-form": "4.20.4",
+    "final-form": "^4.20.4",
     "global": "^4.3.1",
     "history": "^4.10.1",
     "jquery": "^3.6.0",
     "jsonwebtoken": "^9.0.0",
     "localized-strings": "^0.2.4",
     "lodash": "^4.17.21",
     "moment": "^2.29.4",
+    "node-polyfill-webpack-plugin": "^4.0.0",
```
Comment (jennyhliu) [inline package.json:171]: Where is node-polyfill-webpack-plugin used?

File: test/components/executions/execution-events_correctStepName.js
Comment (BWex-NASA) [reply test/components/executions/execution-events_correctStepName.js:1]: We can put them back into one. During conversion it was easier to separate the working portion of the test script with the section that was being troublesome to debug and fix.

Comment (BWex-NASA) [reply test/components/executions/execution-events_correctStepName.js:1]: @jennyhliu I removed the two split out test scripts and converted them back to the one test script: execution.events.js being sure to convert the test code to RTL from Enzyme.

File: CHANGELOG.md (hunk 6 above)
Comment (BWex-NASA) [reply CHANGELOG.md:13]: @jennyhliu Removed the extra line item, as well as the listing of all the unit test scripts that were converted.

File: test/components/granules/bulk-granule-modal.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply test/components/granules/bulk-granule-modal.js:15]: This was agreed to be covered in CUMULUS-3863

File: test/components/collections/list.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply test/components/collections/list.js:61]: @jjmccoy  thanks for clarifying!

File: package.json (hunk 14 above)
Comment (Jkovarik) [reply package.json:108]: I belive it's being used to lint testing-library

File: package.json (hunk 17 above)
Comment (jjmccoy) [reply package.json:171]: I believe it is for the Node.js core modules polyfill in Webpack since it was automatically dropped but not sure which browser polyfill we are still supporting.

File: package.json (hunk 15 above)
Comment (jjmccoy) [reply package.json:147]: I concur we use `react-table`. I'm not familiar with this react-table dependency either.

File: package.json (hunk 14 above)
Comment (jjmccoy) [reply package.json:108]: Yes, it's React Testing Library (RTL) eslint suggested linter dependency.

File: package.json (hunk 13 above)
Comment (jjmccoy) [reply package.json:74]: Yeah, we haven't moved to Jest yet. We are still Ava for unit tests. I think it may have been one of those get all the dependencies when you at RTL and this wasn't noticed when that happened.

File: package.json (hunk 12 above)
Comment (jjmccoy) [reply package.json:148]: I concur. Yeah, only needed @testing-library/react (see our Icomment below about Jest) and @testing-library/dom when installing RTL.

File: localdashboard_cumulus_std.bash
Comment (jjmccoy) [reply localdashboard_cumulus_std.bash:1]: Yeah, we want to delete that instead of doing a `.gitignore` as well.

File: CHANGELOG.md (hunk 7 above)
Comment (Jkovarik) [reply CHANGELOG.md]: 5d3df842

File: CHANGELOG.md (hunk 9 above)
Comment (Jkovarik) [reply CHANGELOG.md]: 174a3ab6

File: package.json (hunk 12 above)
Comment (Jkovarik) [reply package.json]: Agree - updated 0d9e6c39

File: test/components/table/TableFilters.js (hunk 1 above)
Comment (Jkovarik) [reply test/components/table/TableFilters.js:34]: Agreed to handle this in a ticket.   Resolving in PR

File: app/src/js/components/Granules/test.container.html
Comment (Jkovarik) [reply app/src/js/components/Granules/test.container.html:1]: I've removed it as part of the amendment PR

File: package.json (hunk 13 above)
Comment (Jkovarik) [reply package.json]: I removed as part of the amend PR

File: package.json (hunk 16 above)
Comment (Jkovarik) [reply package.json]: Removed in the amend PR

File: package.json (hunk 17 above)
Comment (Jkovarik) [reply package.json]: Agree - I left it in and moved it to the dev dependencies.

**Review by Jkovarik — APPROVED** [review]
With the merged amendments without further changes this looks good!     Great work everyone.

**Review by BWex-NASA — APPROVED** [review]
Approved latest changes from @Jkovarik and @vindigo.

File: CHANGELOG.md [hunk 18]
```diff
@@ -6,11 +6,18 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Changed
 
-This version of the dashboard requires Cumulus API >= v19.2.0-alpha.1 (TBD API release)
```
Comment (jennyhliu) [inline CHANGELOG.md:10]: Keep this line "This version of the dashboard..." under `## [Unreleased]`, it will be updated when we have a core release.
Other than this, my comments are addressed. Thanks.

Comment (BWex-NASA) [reply CHANGELOG.md]: I added it back in. Thanks @jennyhliu.
